src/cryptonote_core/quorum_cop.cpp

```diff
@@ -0,0 +1,221 @@
+        if (!m_core.add_deregister_vote(vote, vvc))
+        {
+          if (vvc.m_invalid_block_height)
+            LOG_ERROR("block height was invalid: " << vote.block_height);
+
+          if (vvc.m_voters_quorum_index_out_of_bounds)
+            LOG_ERROR("voters quorum index specified out of bounds: " << vote.voters_quorum_index);
+
+          if (vvc.m_service_node_index_out_of_bounds)
+            LOG_ERROR("service node index specified out of bounds: " << vote.service_node_index);
+
+          if (vvc.m_signature_not_valid)
+            LOG_ERROR("signature was not valid, was the signature signed properly?");
+        }
+      }
+    }
+
+    prune_uptime_proof_below(execute_justice_from_height);
+  }
+
+  static crypto::hash make_hash(crypto::public_key const &pubkey, uint64_t timestamp)
+  {
+    // TODO(doyle): Errrr clarify what this is. Looks like crypto::cn_fast_hash should be sizeof(buf)
+    char buf[44] = "SUP"; // Meaningless magic bytes
+    crypto::hash result;
+    memcpy(buf + 4, reinterpret_cast<const void *>(&pubkey), sizeof(pubkey));
+    memcpy(buf + 4 + sizeof(pubkey), reinterpret_cast<const void *>(&timestamp), sizeof(timestamp));
+    crypto::cn_fast_hash(buf, sizeof(buf), result);
+
+    return result;
+  }
+
+  bool quorum_cop::handle_uptime_proof(uint64_t timestamp, const crypto::public_key& pubkey, const crypto::signature& sig)
+  {
+    uint64_t now = time(nullptr);
+
+    if ((timestamp < now - UPTIME_PROOF_BUFFER_IN_SECONDS) || (timestamp > now + UPTIME_PROOF_BUFFER_IN_SECONDS))
+      return false;
+
+    if (!m_service_node_list.is_service_node(pubkey))
+      return false;
+
+    CRITICAL_REGION_LOCAL(m_lock);
+    if (m_uptime_proof_seen[pubkey] > now - UPTIME_PROOF_MAX_TIME_IN_SECONDS)
+      return false; // already received one uptime proof for this node recently.
+
+    crypto::hash hash = make_hash(pubkey, timestamp);
+    if (!crypto::check_signature(hash, pubkey, sig))
+      return false;
+
+    m_uptime_proof_seen[pubkey] = timestamp;
+    return true;
+  }
+
+  void quorum_cop::generate_uptime_proof_request(const crypto::public_key& pubkey, const crypto::secret_key& seckey, cryptonote::NOTIFY_UPTIME_PROOF::request& req) const
+  {
+    req.timestamp     = time(nullptr);
+    req.pubkey        = pubkey;
+
+    crypto::hash hash = make_hash(req.pubkey, req.timestamp);
+    crypto::generate_signature(hash, pubkey, seckey, req.sig);
+  }
+
+  void quorum_cop::prune_uptime_proof_below(uint64_t height)
+  {
+    std::list<cryptonote::block> blocks;
+    if (!m_core.get_blocks(height, 1, blocks))
+    {
+      LOG_ERROR("Error quorum cop could not retrieve block with height: " << height << ", from db");
+      return;
+    }
+
+    const uint64_t prune_from_timestamp = blocks.front().timestamp;
+    CRITICAL_REGION_LOCAL(m_lock);
+
+    for (auto it = m_uptime_proof_seen.begin(); it != m_uptime_proof_seen.end();)
+      it = (it->second < prune_from_timestamp) ? m_uptime_proof_seen.erase(it) : it;
```

Comment:
This is an infinite loop.

change to:

    if (it->second < prune_from_timestamp)
        it = m_uptime_proof_seen.erase(it);
    else
        it++;

Reply:
Alternatively just add ++it to the second branch in your ternary operator. Not sure if this is defined behaviour.